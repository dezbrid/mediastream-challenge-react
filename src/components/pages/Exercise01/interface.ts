export interface Movie {
  id: number;
  name: string;
  price: number;
}
export interface DiscountRules {
  m: number[];
  discount: number;
}
export interface Cart extends Movie {
  quantity: number;
}

type PartialProperty<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>;

export interface MovieCardInterfaces extends PartialProperty<Cart, "quantity"> {
  disableAddtoCart?: boolean;
}
