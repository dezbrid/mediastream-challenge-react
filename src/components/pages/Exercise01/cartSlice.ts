import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { Cart, Movie } from "./interface";
import { RootState } from "../../../app/store";

export interface CartState {
  listCart: Cart[];
}

const initialState: CartState = {
  listCart: [],
};

export const cartSlice = createSlice({
  name: "cart",
  initialState,
  reducers: {
    addCart: (state, action: PayloadAction<Movie>) => {
      state.listCart = [...state.listCart, { ...action.payload, quantity: 1 }];
    },
    incrementItemByID: (state, action: PayloadAction<number>) => {
      state.listCart = editCart(state.listCart, action.payload, true);
    },
    decrementItemByID: (state, action: PayloadAction<number>) => {
      state.listCart = editCart(state.listCart, action.payload, false);
    },
  },
});

const editCart = (list: Cart[], id: number, plus: boolean): Cart[] => {
  const index = list.findIndex((item) => item.id === id);
  const currentQuantity = list[index].quantity;
  if (currentQuantity === 1 && !plus) {
    return [...list.slice(0, index), ...list.slice(index + 1)];
  }
  return [
    ...list.slice(0, index),
    {
      ...list[index],
      quantity: plus ? currentQuantity + 1 : currentQuantity - 1,
    },
    ...list.slice(index + 1),
  ];
};
export const selectList = (state: RootState) => state.cart.listCart;
export const { addCart, incrementItemByID, decrementItemByID } =
  cartSlice.actions;

export default cartSlice.reducer;
