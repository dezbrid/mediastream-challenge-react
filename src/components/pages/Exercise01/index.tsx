/**
 * Exercise 01: The Retro Movie Store
 * Implement a shopping cart with the next features for the Movie Store that is selling retro dvds:
 * 1. Add a movie to the cart
 * 2. Increment or decrement the quantity of movie copies. If quantity is equal to 0, the movie must be removed from the cart
 * 3. Calculate and show the total cost of your cart. Ex: Total: $150
 * 4. Apply discount rules. You have an array of offers with discounts depending of the combination of movie you have in your cart.
 * You have to apply all discounts in the rules array (discountRules).
 * Ex: If m:Â [1, 2, 3], it means the discount will be applied to the total when the cart has all that products in only.
 *
 * You can modify all the code, this component isn't well designed intentionally. You can redesign it as you need.
 */
import React from "react";
import "./assets/styles.css";

import { useAppSelector } from "../../../app/hooks";

import MovieCard from "./components/MovieCard";
import { selectList } from "./cartSlice";
import { MOVIES, DISCOUNTRULES } from "./mocks";

export default function Exercise01() {
  const listCart = useAppSelector(selectList);

  const iteminList = (id: number): boolean => {
    return listCart.some((item) => item.id === id);
  };
  const handleDicount = ():number => {
    for (let rules of DISCOUNTRULES) {
      const inRules = rules.m.reduce(
        (pre, curr) => pre && iteminList(curr),
        true
      );
      if (inRules) return rules.discount;
    }
    return 0
  };
  const getTotal = () => {
    const subtotal=
    listCart.reduce(
      (previousValue, currentValue) =>
        previousValue + currentValue.price * currentValue.quantity,
      0
    );
    return subtotal- handleDicount()
  }; // TODO: Implement this

  return (
    <section className="exercise01">
      <div className="movies__container movies__container-space">
        <ul>
          {MOVIES.map((movieItem) => (
            <MovieCard
              {...movieItem}
              key={movieItem.id}
              disableAddtoCart={iteminList(movieItem.id)}
            />
          ))}
        </ul>
      </div>
      <div className="movies__container">
        <ul className="movies__container-cart">
          {listCart.map((cartItem) => (
            <MovieCard {...cartItem} key={cartItem.id} />
          ))}
        </ul>
        <div className="movies__cart-total">
          <p>Total: ${getTotal()}</p>
        </div>
      </div>
    </section>
  );
}
