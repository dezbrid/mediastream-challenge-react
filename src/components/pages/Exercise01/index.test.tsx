import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import Exercise01 from "./index";
import { store } from '../../../app/store';

describe("<Exercise01 />", () => {
  it("<Exercise01 /> snapshot", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Exercise01 />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
