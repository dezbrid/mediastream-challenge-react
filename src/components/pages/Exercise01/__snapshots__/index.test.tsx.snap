// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Exercise01 /> <Exercise01 /> snapshot 1`] = `
<section
  className="exercise01"
>
  <div
    className="movies__container movies__container-space"
  >
    <ul>
      <li
        className="movies__card"
      >
        <ul>
          <li>
            ID:
            1
          </li>
          <li>
            Name: 
            Star Wars
          </li>
          <li>
            Price: $
            20
          </li>
        </ul>
        <button
          className="movies__card-add-cart"
          disabled={false}
          onClick={[Function]}
        >
          Add to cart
        </button>
      </li>
      <li
        className="movies__card"
      >
        <ul>
          <li>
            ID:
            2
          </li>
          <li>
            Name: 
            Minions
          </li>
          <li>
            Price: $
            25
          </li>
        </ul>
        <button
          className="movies__card-add-cart"
          disabled={false}
          onClick={[Function]}
        >
          Add to cart
        </button>
      </li>
      <li
        className="movies__card"
      >
        <ul>
          <li>
            ID:
            3
          </li>
          <li>
            Name: 
            Fast and Furious
          </li>
          <li>
            Price: $
            10
          </li>
        </ul>
        <button
          className="movies__card-add-cart"
          disabled={false}
          onClick={[Function]}
        >
          Add to cart
        </button>
      </li>
      <li
        className="movies__card"
      >
        <ul>
          <li>
            ID:
            4
          </li>
          <li>
            Name: 
            The Lord of the Rings
          </li>
          <li>
            Price: $
            5
          </li>
        </ul>
        <button
          className="movies__card-add-cart"
          disabled={false}
          onClick={[Function]}
        >
          Add to cart
        </button>
      </li>
    </ul>
  </div>
  <div
    className="movies__container"
  >
    <ul
      className="movies__container-cart"
    />
    <div
      className="movies__cart-total"
    >
      <p>
        Total: $
        0
      </p>
    </div>
  </div>
</section>
`;
